pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,390 @@
+		log:              c.Logger.WithField("subsystem_name", "cacheManager"),
+		spiffeIdEntryMap: make(map[string]CacheEntry),
+		entryRequestCh:   make(chan map[string][]EntryRequest),
+		regEntriesCh:     make(chan []*proto.RegistrationEntry),
+		cacheEntryCh:     make(chan CacheEntry),
+		ctx:              ctx,
+		cancel:           cancel,
+		doneCh:           make(chan struct{}),
+		errch:            make(chan error),
+	}, nil
+}
+
+func (m *manager) Shutdown(err error) {
+	m.reason = err
+	m.cancel()
+}
+
+func (m *manager) Done() <-chan struct{} {
+	return m.doneCh
+}
+func (m *manager) Err() error {
+	<-m.Done()
+	return m.reason
+}
+func (m *manager) Cache() Cache {
+	return m.managedCache
+}
+
+func (m *manager) Init() {
+	go func() {
+		defer close(m.doneCh)
+		defer close(m.errch)
+		var wg sync.WaitGroup
+
+		wg.Add(1)
+		go m.expiredCacheEntryHandler(5*time.Second, &wg)
+
+		wg.Add(1)
+		go m.regEntriesHandler(&wg)
+		m.log.Debug("Initializing Cache Manager")
+		m.regEntriesCh <- m.baseRegEntries
+		for {
+
+			var svid []byte
+			var key *ecdsa.PrivateKey
+			select {
+			case reqs := <-m.entryRequestCh:
+				for parentId, entryRequests := range reqs {
+					wg.Add(1)
+					if _, ok := m.spiffeIdEntryMap[parentId]; ok {
+						svid = m.spiffeIdEntryMap[parentId].SVID.SvidCert
+						key = m.spiffeIdEntryMap[parentId].PrivateKey
+					}
+					if parentId == m.baseSPIFFEID {
+						svid = m.baseSVID
+						key = m.baseSVIDKey
+					}
+					conn, err := m.getGRPCConn(svid, key)
+					if err != nil {
+						m.errch <- err
+						break
+					}
+					go m.fetchSVID(entryRequests, node.NewNodeClient(conn), &wg)
+				}
+
+			case newCacheEntry := <-m.cacheEntryCh:
+				m.managedCache.SetEntry(newCacheEntry)
+				m.log.Debugf("Updated CacheEntry: %s", newCacheEntry)
+
+			case <-m.ctx.Done():
+				wg.Wait()
+				return
+
+			case err := <-m.errch:
+				m.Shutdown(err)
+				for err := range m.errch {
+					m.log.Debug(err)
+				}
+
+			}
+		}
+	}()
+}
+
+func (m *manager) fetchSVID(requests []EntryRequest, nodeClient node.NodeClient, wg *sync.WaitGroup) {
+	defer wg.Done()
+	stream, err := nodeClient.FetchSVID(context.Background())
+	if err != nil {
+		m.errch <- err
+		return
+	}
+	defer func() {
+		if err := stream.CloseSend(); err != nil {
+			m.errch <- err
+			return
+		}
+	}()
+
+	for _, req := range requests {
+
+		err := stream.Send(&node.FetchSVIDRequest{Csrs: append([][]byte{}, req.CSR)})
+
+		if err != nil {
+			m.errch <- err
+			return
+		}
+
+		resp, err := stream.Recv()
+		if err != nil {
+			m.errch <- err
+			return
+		}
+		svid := resp.SvidUpdate.Svids[req.entry.RegistrationEntry.SpiffeId]
+		cert, err := x509.ParseCertificate(svid.SvidCert)
+		if err != nil {
+			m.errch <- err
+			return
+		}
+
+		m.regEntriesCh <- resp.SvidUpdate.RegistrationEntries
+		req.entry.SVID = svid
+		req.entry.Expiry = cert.NotAfter
+		select {
+		case m.cacheEntryCh <- req.entry:
+		case <-m.ctx.Done():
+			m.errch <- m.ctx.Err()
```

Comment:
deadlock.

Reply:
@boz would it make sense  wait on the `m.errch` in a separate `go func()` to avoid these deadlocks:
```	
	go func() {
		defer close(m.errch)
		err := <-m.errch
		m.Shutdown(err)
		for err := range m.errch {
			m.log.Debug(err)
		}}()
```

Reply:
No.

 * what if nothing is ever sent to `m.errch`?
 * `m.errch` is not closed before the `for` loop so that will block forever.

Don't use `m.errch`.  Use `Shutdown(error)`.  Log in that method.  Populate `m.reason` with the error that gets passed the first time `Shutdown()` is called using `sync.Once` like I mentioned earlier.

Alternatively, look at [tomb](https://godoc.org/gopkg.in/tomb.v2#Tomb.Kill) or [go-lifecycle](https://github.com/boz/go-lifecycle) for other pattern options.

Reply:
ok let me go back to using `Shutdown(error) with `sync.Once`. Thank you.